lib/ansible/modules/cloud/amazon/ec2_metadata_facts.py

```diff
@@ -445,7 +446,16 @@ def __init__(self, module, ec2_metadata_uri=None, ec2_sshdata_uri=None, ec2_user
         self._prefix = 'ansible_ec2_%s'
 
     def _fetch(self, url):
-        (response, info) = fetch_url(self.module, url, force=True)
+        response, info = fetch_url(self.module, url, force=True)
+
+        if not info.get('msg', '').startswith('OK'):
+            time.sleep(3)
+            # request went bad, retry once then raise
+            self.module.warn('Retrying query to metadata service. First attempt failed: {0}'.format(info['msg']))
+            response, info = fetch_url(self.module, url, force=True)
+            if not info.get('msg', '').startswith('OK'):
+                # fail out now
+                self.module.fail_json(msg='Failed to retrieve metadata from AWS: {0}'.format(info['msg']), response=info)
         if response:
             data = response.read()
         else:
```

Comment:
Should this if-else block be changed to be just an else block? Or do we still want to return empty data when there is no response?

Reply:
Hm. I suppose in the case where nothing is returned we should probably fail out entirely? Or are there cases where it's intentional that the metadata service returns no body?

Reply:
Ah never mind, I forgot that the `if response` handles when the recursive URL lookup might have ended.